reviewing/dependencies-hierarchy/src/TreeNodeId.ts

```diff
@@ -0,0 +1,26 @@
+export type TreeNodeId = TreeNodeIdImporter | TreeNodeIdPackage
+
+/**
+ * A project local to the pnpm workspace.
+ */
+interface TreeNodeIdImporter {
+  readonly type: 'importer'
+  readonly importerId: string
+}
+
+/**
+ * An npm package depended on externally.
+ */
+interface TreeNodeIdPackage {
+  readonly type: 'package'
+  readonly depPath: string
+}
+
+export function serializeTreeNodeId (treeNodeId: TreeNodeId): string {
+  switch (treeNodeId.type) {
+  case 'importer':
+    return `importer__${treeNodeId.importerId}`
+  case 'package':
+    return `package__${treeNodeId.depPath}`
+  }
```

Comment:
Why don't you just do `JSON.stringify(treeNodeId)`?

Reply:
Since TypeScript is duck-typed, I try to defend against extra unknown props when serializing.

```ts
const importerA = {
  type: 'importer',
  importerId: '...',
}

const importerB = {
  ...importerA,
  extraUnknownField: '...',
}
```

In this case:
```ts
isNodeIdEqual(importerA, importerB) === true
```

But if we were to serialize the entire object:

```ts
serializeTreeNodeId(importerA) !== serializeTreeNodeId(importerB)
```

How do you feel about this?

```ts
export function serializeTreeNodeId (treeNodeId: TreeNodeId): string {
  switch (treeNodeId.type) {
  case 'importer': {
    // Only serialize known fields.
    const { type, importerId } = treeNodeId
    return JSON.stringify({ type, importerId })
  }
  case 'package': {
    const { type, depPath } = treeNodeId
    return JSON.stringify({ type, depPath })
  }
  }
}
```

Reply:
Latest push contains the code above, but let me know if you'd prefer to just straight `JSON.stringify`.